pingcap/tidb
# expression, planner: support builtin function benchmark

### What problem does this PR solve? 

Implements builtin function `BENCHMARK()`, and introduces a counter in expression_rewriter to disable constant folding. (#6774)

### What is changed and how it works?

1. Implements builtin `BENCHMARK()` function.
1. To disable constant folding inside `BENCHMARK()` scope, `expressionRewriter.disableFoldCounter` is introduced.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test
```
select benchmark(1000000000, 3);
select benchmark(1000000000, length("abc"));
select benchmark(1000000000, sin(123) + sin(456));
```
Execurtion time of these 3 commands should differ.

Code changes

 - `builtinBenchmarkSig` in **expression/builtin_info.go**
 - `expressionRewriter.disableFoldCounter` in **planner/core/expression_rewriter.go**
 - `expressionRewriter.newFunction()` in **planner/core/expression_rewriter.go** to control new function behavior by the counter.

Side effects

 - Everytime the expression_rewriter wants to get a new builtin function expression, it has to use `expressionRewriter.newFunction()` instead of `expression.NewFunction()`, or this builtin function may not be available to be measured by `BENCHMARK()` because of constant folding.

Related changes

 - Nothing.
